modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/TabsTag.java

```diff
@@ -0,0 +1,189 @@
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.taglib.clay.servlet.taglib;
+
+import com.liferay.frontend.taglib.clay.internal.servlet.taglib.BaseContainerTag;
+import com.liferay.frontend.taglib.clay.servlet.taglib.util.TabsItem;
+import com.liferay.portal.kernel.language.LanguageUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.taglib.util.TagResourceBundleUtil;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.JspWriter;
+
+/**
+ * @author Carlos Lancha
+ */
+public class TabsTag extends BaseContainerTag {
+
+	@Override
+	public int doStartTag() throws JspException {
+		setAttributeNamespace(_ATTRIBUTE_NAMESPACE);
+
+		return super.doStartTag();
+	}
+
+	public String getActivation() {
+		return _activation;
+	}
+
+	public String getDisplayType() {
+		return _displayType;
+	}
+
+	public int getPanelsCount() {
+		return _panelsCount;
+	}
+
+	public List<TabsItem> getTabsItems() {
+		return _tabsItems;
+	}
+
+	public boolean isFade() {
+		return _fade;
+	}
+
+	public boolean isJustified() {
+		return _justified;
+	}
+
+	public void setActivation(String activation) {
+		_activation = activation;
+	}
+
+	public void setDisplayType(String displayType) {
+		_displayType = displayType;
+	}
+
+	public void setFade(boolean fade) {
+		_fade = fade;
+	}
+
+	public void setJustified(boolean justified) {
+		_justified = justified;
+	}
+
+	public void setPanelsCount(int panelsCount) {
+		_panelsCount = panelsCount;
+	}
+
+	public void setTabsItems(List<TabsItem> tabsItems) {
+		_tabsItems = tabsItems;
+	}
+
+	@Override
+	protected void cleanUp() {
+		super.cleanUp();
+
+		_activation = "manual";
+		_displayType = null;
+		_fade = false;
+		_justified = false;
+		_panelsCount = 0;
+		_tabsItems = null;
+	}
+
+	@Override
+	protected String getHydratedModuleName() {
+		return "{Tabs} from frontend-taglib-clay";
+	}
+
+	@Override
+	protected Map<String, Object> prepareProps(Map<String, Object> props) {
+		props.put("activation", _activation);
+		props.put("displayType", _displayType);
+		props.put("fade", _fade);
+		props.put("justified", _justified);
+		props.put("tabsItems", _tabsItems);
+
+		return super.prepareProps(props);
+	}
+
+	@Override
+	protected void processStartBodyTag() throws Exception {
+		JspWriter jspWriter = pageContext.getOut();
+
+		jspWriter.write("<div class=\"tab-content tag-body-content\">");
```

Comment:
I'm ok with forcing each composable tag to add the `tag-body-content` marker class manually, however, it's not clear to me whether this is the sole purpose of `processStartBodyTag()` method or there may be other things we'd like to do there. 

In this case it seems we need to add the `tab-content` class, but maybe this can be handled with the `setCssClass()` instead?


If there are not, then we should delegate this to the `BaseContainerTag` because it has the required information to know about. This way, composable tags with hydration would be automaticallty recognized by the infra as long as:
* They return `EVAL_BODY_INCLUDE` (i.e. they're composable)
* They have hydration in the current instance render cycle

Looks like this would be enough for the other tags (NavigationBar, Link, Button) which would benefit from this for free

Reply:
Same as before, `setCssClass` or `cssClass.add` acts over the main element, which is not the one we need to add the class.

The rendering of `tab-body-content` [is already a reponsibility of `BaseContainerTag`](https://github.com/liferay-frontend/liferay-portal/pull/3045/files#diff-2b4cdf2f6f59a9e1dcf930ab065f8d291606569ee9732953b52b52bff5b941eaR417), you only need to override the method `processStartBodyTag` if you need to render something different like is the case of Tabs, where we need that div to also have `tab-content` class. If you don't need to modify it you don't need to implement the method, so other taglibs can already benefit from it for free.

Reply:
What we can add is a method `processBodyCssClass` so if you only need to add classes to the body container you implement that class instead of having to implement `processStartBodyTag` and we'd be mimicking the main element behavior.

Or even easier, pass to `processCssClass` a new argument called `bodyCssClass` so you can add in that method both cssClasses for the container and cssClasses for the body Container.

```
protected String processCssClasses(Set<String> cssClasses, Set<String> bodyCssClasses) {
		cssClasses.add("main-container-extra-class");

		bodyCssClasses.add("body-container-extra-class");

		return super.processCssClasses(cssClasses, bodyCssClasses);
	}
```

Reply:
Forget about the last option, `processCssClasses` return is directly rendered as the container Css Classes, is better to just create a new `processBodyCssClasses` method.